src/Analyser/MutatingScope.php

```diff
@@ -909,12 +909,17 @@ private function resolveType(Expr $node): Type
 			} else {
 				$expressionType = $this->getNativeType($node->expr);
 			}
-			if (
-				$this->isInTrait()
-				&& TypeUtils::findThisType($expressionType) !== null
-			) {
-				return new BooleanType();
+
+			if ($this->isInTrait()) {
+				if ($node->expr instanceof Expr\Variable
```

Comment:
when inside a trait, it seems `$node->expr` is resolved into a `NeverType`.

actually I thought it should return a `ThisType` and it should be catched by `TypeUtils::findThisType` below, but it doesn't work like that.

Reply:
looked into it again and I am still not know more then the comment above.

maybe it even intentionally resolves a 'this' variable to NeverType within a trait, as technically the trait only has a 'this' when used in a class context?

Reply:
ohh I got an idea. I guess the cause is that we are trying to remove `ObjectType` from `ThisType`.